From f683ab785704cd9de36c2d929a45604a9c9feb3f Mon Sep 17 00:00:00 2001
From: Aloka Dixit <quic_alokad@quicinc.com>
Date: Mon, 28 Mar 2022 14:32:19 -0700
Subject: [PATCH] ath12k: skip sending vdev down for channel switch

Firmware crash is seen during channel switch when multiple BSSID is
enabled. This is due to ath12k driver calling ath12k_wmi_vdev_down()
before ath11k_mac_vdev_restart(). As firmware brings down the vdevs
internally as part of restart already, remove the unnecessary
operation to fix the crash.

Signed-off-by: Aloka Dixit <quic_alokad@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 33 +++++----------------------
 1 file changed, 6 insertions(+), 27 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 7f8a7b4..1fea8d5 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -7314,7 +7314,11 @@ ath12k_mac_update_vif_chan(struct ath12k *ar,
 
 	lockdep_assert_held(&ar->conf_mutex);
 
+	/* TODO: Update ar->rx_channel */
+
 	for (i = 0; i < n_vifs; i++) {
+		struct vdev_up_params params = { 0 };
+
 		arvif = (void *)vifs[i].vif->drv_priv;
 
 		if (vifs[i].vif->type == NL80211_IFTYPE_MONITOR)
@@ -7331,41 +7335,16 @@ ath12k_mac_update_vif_chan(struct ath12k *ar,
 		if (WARN_ON(!arvif->is_started))
 			continue;
 
-		if (WARN_ON(!arvif->is_up))
-			continue;
-
-		ret = ath12k_wmi_vdev_down(ar, arvif->vdev_id);
+		ret = ath12k_mac_vdev_restart(arvif, &vifs[i].new_ctx->def);
 		if (ret) {
-			ath12k_warn(ab, "failed to down vdev %d: %d\n",
+			ath12k_warn(ab, "failed to restart vdev %d: %d\n",
 				    arvif->vdev_id, ret);
 			continue;
 		}
-	}
-
-	/* All relevant vdevs are downed and associated channel resources
-	 * should be available for the channel switch now.
-	 */
-
-	/* TODO: Update ar->rx_channel */
-
-	for (i = 0; i < n_vifs; i++) {
-		struct vdev_up_params params = { 0 };
-
-		arvif = (void *)vifs[i].vif->drv_priv;
-
-		if (WARN_ON(!arvif->is_started))
-			continue;
 
 		if (WARN_ON(!arvif->is_up))
 			continue;
 
-		ret = ath12k_mac_vdev_restart(arvif, &vifs[i].new_ctx->def);
-		if (ret) {
-			ath12k_warn(ab, "failed to restart vdev %d: %d\n",
-				    arvif->vdev_id, ret);
-			continue;
-		}
-
 		ret = ath12k_mac_setup_bcn_tmpl(arvif);
 		if (ret)
 			ath12k_warn(ab, "failed to update bcn tmpl during csa: %d\n",
-- 
2.17.1

